pkg/schedule/windowed_schedule.go

```diff
@@ -12,35 +31,72 @@ var (
 
 // WindowedSchedule is a schedule that waits on an interval within a specific time window
 type WindowedSchedule struct {
-	Interval  time.Duration
-	StartTime *time.Time
-	StopTime  *time.Time
-	state     ScheduleState
+	Interval   time.Duration
+	StartTime  *time.Time
+	StopTime   *time.Time
+	Count      uint
+	state      ScheduleState
+	stopOnTime *time.Time
 }
 
-// NewWindowedSchedule returns an instance of WindowedSchedule given duration,
-// start and stop time
-func NewWindowedSchedule(i time.Duration, start *time.Time, stop *time.Time) *WindowedSchedule {
+// NewWindowedSchedule returns an instance of WindowedSchedule with given interval, start and stop timestamp
+// and count of expected runs. The value of `count` determines stop time, so specifying it together with `stop`
+// is not allowed and the count will be set to defaults 0 in such cases.
+func NewWindowedSchedule(i time.Duration, start *time.Time, stop *time.Time, count uint) *WindowedSchedule {
+	// if stop and count were both defined, ignore the `count`
+	if count != 0 && stop != nil {
+		count = 0
+		// log about ignoring the `count`
+		logger.WithFields(log.Fields{
+			"_block": "NewWindowedSchedule",
+		}).Warning("Specifying both the window stop and the count is not allowed. The parameter `count` has been ignored.")
+	}
+
 	return &WindowedSchedule{
 		Interval:  i,
 		StartTime: start,
 		StopTime:  stop,
+		Count:     count,
 	}
 }
 
+// setStopOnTime calculates and set the value of the windowed `stopOnTime` which is the right window boundary.
+// `stopOnTime` is determined by `StopTime` or, if it is not provided, calculated based on count and interval.
+func (w *WindowedSchedule) setStopOnTime() {
+	if w.StopTime == nil && w.Count != 0 {
+		// determine the window stop based on the `count` and `interval`
+		var newStop time.Time
+
+		// if start is not set or points in the past,
+		// use the current time to calculate stopOnTime
+		if w.StartTime != nil && time.Now().Before(*w.StartTime) {
+			newStop = w.StartTime.Add(time.Duration(w.Count) * w.Interval)
+		} else {
+			// set a new stop timestamp from this point in time
+			newStop = time.Now().Add(time.Duration(w.Count) * w.Interval)
+		}
```

Comment:
yes, this is covered in [medium tests] (https://github.com/IzabellaRaulin/snap/blob/run_task_x_times/pkg/schedule/windowed_schedule_medium_test.go) - this function is called on each first "schedule.Wait", so all tests cases apply to that, but especially those ones [here](https://github.com/IzabellaRaulin/snap/blob/run_task_x_times/pkg/schedule/windowed_schedule_medium_test.go#L332)